devito/interpolators.py

```diff
@@ -0,0 +1,410 @@
+import sympy
+import numpy as np
+
+from abc import ABC, abstractmethod
+from cached_property import cached_property
+
+from devito.logger import warning
+from devito.symbolics import retrieve_function_carriers, indexify, INT
+from devito.tools import Evaluable, powerset, flatten, prod
+from devito.types import Eq, Inc
+from devito.types.basic import Scalar
+from devito.types.dense import SubFunction
+from devito.types.dimension import ConditionalDimension, Dimension, DefaultDimension
+
+
+class UnevaluatedSparseOperation(Evaluable):
+
+    """
+    Represents an Injection or an Interpolation operation performed on a
+    SparseFunction. Evaluates to a list of Eq objects.
+
+    Parameters
+    ----------
+    interpolator : Interpolator
+        Interpolator object that will be used to evaluate the Operation.
+    *args, **kwargs
+        The arguments passed to the corresponding method.
+    """
+    subdomain = None
+
+    def __init__(self, interpolator, *args, **kwargs):
+        assert(isinstance(interpolator, GenericInterpolator))
+
+        self.interpolator = interpolator
+        self._args = args
+        self._kwargs = kwargs
+
+    def __add__(self, other):
+        return flatten([self, other])
+
+    def __radd__(self, other):
+        return flatten([other, self])
+
+
+class Injection(UnevaluatedSparseOperation):
+
+    """
+    Represents an Injection operation performed on a SparseFunction.
+    Evaluates to a list of Eq objects.
+    """
+
+    def __init__(self, interpolator, field, expr, *args, **kwargs):
+        self.field = field
+        self.expr = expr
```

Comment:
Not used anywhere, why is it saved here.

Reply:
It is used in the repr. This is part of the code I was forced to add because we wanted pretty printing of these classes.